src/core/public/apm_system.ts

```diff
@@ -64,11 +68,25 @@ export class ApmSystem {
 
     this.markPageLoadStart();
 
+    start.executionContext.context$.subscribe((c) => {
+      // We're using labels because we want the context to be indexed
+      // https://www.elastic.co/guide/en/apm/get-started/current/metadata.html
+      const apmContext = {
+        appId: c.name,
+        page: c.page,
+        id: c.id,
+      };
+
+      this.apm?.addLabels(apmContext);
+    });
+
+    // TODO: Start a new transaction every page change instead of every app change.
+
     /**
      * Register listeners for navigation changes and capture them as
      * route-change transactions after Kibana app is bootstrapped
      */
-    start.application.currentAppId$.subscribe((appId) => {
+    start.application.currentAppId$.pipe(distinctUntilChanged()).subscribe((appId) => {
```

Comment:
NIT: I think this is unnecessary, as we're already using this operator on the exposed observable.

https://github.com/elastic/kibana/blob/d364f237c5dd2309eb7e61c3c46399048719fa25/src/core/public/application/application_service.tsx#L277-L281